src/OrchardCore.Modules/OrchardCore.Users/Services/DefaultControllerService.cs

```diff
@@ -0,0 +1,133 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.Extensions.Logging;
+using OrchardCore.DisplayManagement;
+using OrchardCore.Email;
+using OrchardCore.Entities;
+using OrchardCore.Modules;
+using OrchardCore.Settings;
+using OrchardCore.Users.Events;
+using OrchardCore.Users.Models;
+using OrchardCore.Users.ViewModels;
+
+namespace OrchardCore.Users.Services
+{
+    public class DefaultControllerService
+    {
+        private readonly ISmtpService _smtpService;
+        private readonly IDisplayHelper _displayHelper;
+        private readonly HtmlEncoder _htmlEncoder;
+        private readonly IEnumerable<IRegistrationFormEvents> _registrationFormEvents;
+        private readonly IUserService _userService;
+        private readonly ISiteService _siteService;
+        private readonly SignInManager<IUser> _signInManager;
+        private readonly UserManager<IUser> _userManager;
+        private readonly ILogger<DefaultControllerService> _logger;
+
+        public DefaultControllerService(
+            ISmtpService smtpService,
+            IDisplayHelper displayHelper,
+            HtmlEncoder htmlEncoder,
+            IEnumerable<IRegistrationFormEvents> registrationFormEvents,
+            IUserService userService,
+            ISiteService siteService,
+            SignInManager<IUser> signInManager,
+            UserManager<IUser> userManager,
+            ILogger<DefaultControllerService> logger)
+        {
+            _smtpService = smtpService;
+            _displayHelper = displayHelper;
+            _htmlEncoder = htmlEncoder;
+            _registrationFormEvents = registrationFormEvents;
+            _userService = userService;
+            _siteService = siteService;
+            _signInManager = signInManager;
+            _userManager = userManager;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Returns the created user, otherwise returns null
+        /// </summary>
+        /// <param name="controller"></param>
+        /// <param name="model"></param>
+        /// <param name="confirmationEmailSubject"></param>
+        /// <returns></returns>
+        public virtual async Task<IUser> RegisterUser(Controller controller, RegisterViewModel model, string confirmationEmailSubject)
```

Comment:
I would suggest making a context class with an `IUrlHelper` rather than passing the controller class in directly.

I think this should all just go into the `IUserService`

@Skrypt the previous was internal, so no need to deprecate it, as it is unusable by anyone

Reply:
> I would suggest making a context class with an IUrlHelper rather than passing the controller class in directly.

I started this PR by using `LinkGenerator` instead, but I don't want to lose any data from the controller http context, I will try to use `IUrlHelper`

> @Skrypt the previous was internal, so no need to deprecate it, as it is unusable by anyone

You are right